File: sample/java/11_lissajous_particles.java
final class _11_lissajous_particles {
    private _11_lissajous_particles() {
    }


    // 11: Sample that outputs Lissajous-motion particles as a GIF.

    public static java.util.ArrayList<Long> color_palette() {
        java.util.ArrayList<Long> p = new java.util.ArrayList<Long>();
        for (long i = 0L; i < 256L; i += 1L) {
            long r = i;
            long g = i * 3L % 256L;
            long b = 255L - i;
            p.add(r);
            p.add(g);
            p.add(b);
        }
        return PyRuntime.__pytra_bytearray(p);
    }

    public static void run_11_lissajous_particles() {
        long w = 320L;
        long h = 240L;
        long frames_n = 360L;
        long particles = 48L;
        String out_path = "sample/out/11_lissajous_particles.gif";
        double start = time.perf_counter();
        java.util.ArrayList<java.util.ArrayList<Long>> frames = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long t = 0L; t < frames_n; t += 1L) {
            java.util.ArrayList<Long> frame = PyRuntime.__pytra_bytearray(w * h);
            for (long p = 0L; p < particles; p += 1L) {
                double phase = ((double)(p)) * 0.261799;
                long x = PyRuntime.__pytra_int(((double)(w)) * 0.5 + ((double)(w)) * 0.38 * math.sin(0.11 * ((double)(t)) + phase * 2.0));
                long y = PyRuntime.__pytra_int(((double)(h)) * 0.5 + ((double)(h)) * 0.38 * math.sin(0.17 * ((double)(t)) + phase * 3.0));
                long color = 30L + p * 9L % 220L;
                for (long dy = (-(2L)); dy < 3L; dy += 1L) {
                    for (long dx = (-(2L)); dx < 3L; dx += 1L) {
                        long xx = x + dx;
                        long yy = y + dy;
                        if ((((xx) >= (0L)) && ((xx) < (w)) && ((yy) >= (0L)) && ((yy) < (h)))) {
                            long d2 = dx * dx + dy * dy;
                            if (((d2) <= (4L))) {
                                long idx = yy * w + xx;
                                long v = color - d2 * 20L;
                                v = Math.max(0L, v);
                                if (((v) > (((Long)(frame.get((int)((((idx) < 0L) ? (((long)(frame.size())) + (idx)) : (idx))))))))) {
                                    frame.set((int)((((idx) < 0L) ? (((long)(frame.size())) + (idx)) : (idx))), v);
                                }
                            }
                        }
                    }
                }
            }
            frames.add(PyRuntime.__pytra_bytearray(frame));
        }
        gif.save_gif(out_path, w, h, frames, color_palette(), 3L, 0L);
        double elapsed = time.perf_counter() - start;
        System.out.println(String.valueOf("output:") + " " + String.valueOf(out_path));
        System.out.println(String.valueOf("frames:") + " " + String.valueOf(frames_n));
        System.out.println(String.valueOf("elapsed_sec:") + " " + String.valueOf(elapsed));
    }
}

File: sample/julia/11_lissajous_particles.jl
include(joinpath(@__DIR__, "built_in", "py_runtime.jl"))

include(joinpath(@__DIR__, "std", "math.jl"))
math = (ceil=ceil, cos=cos, e=e, exp=exp, fabs=fabs, floor=floor, log=log, log10=log10, pi=pi, pow=pow, sin=sin, sqrt=sqrt, tan=tan)
include(joinpath(@__DIR__, "std", "time.jl"))
include(joinpath(@__DIR__, "utils", "gif.jl"))

# 11: Sample that outputs Lissajous-motion particles as a GIF.

function color_palette()
    p = UInt8[]
    for i in 0:256 - 1
        r = i
        g = ((i * 3) % 256)
        b = (255 - i)
        push!(p, r)
        push!(p, g)
        push!(p, b)
    end
    return __pytra_bytes(p)
end

function run_11_lissajous_particles()
    w = 320
    h = 240
    frames_n = 360
    particles = 48
    out_path = "sample/out/11_lissajous_particles.gif"
    
    start = perf_counter()
    frames = Any[]
    
    for t in 0:frames_n - 1
        frame = __pytra_bytearray((w * h))
        
        for p in 0:particles - 1
            phase = (p * 0.261799)
            x = __pytra_int(((w * 0.5) + ((w * 0.38) * math.sin(((0.11 * t) + (phase * 2.0))))))
            y = __pytra_int(((h * 0.5) + ((h * 0.38) * math.sin(((0.17 * t) + (phase * 3.0))))))
            color = (30 + ((p * 9) % 220))
            
            for dy in (-2):3 - 1
                for dx in (-2):3 - 1
                    xx = (x + dx)
                    yy = (y + dy)
                    if ((xx >= 0) && (xx < w) && (yy >= 0) && (yy < h))
                        d2 = ((dx * dx) + (dy * dy))
                        if (d2 <= 4)
                            idx = ((yy * w) + xx)
                            v = (color - (d2 * 20))
                            v = max(0, v)
                            if (v > frame[__pytra_idx(idx, length(frame))])
                                frame[__pytra_idx(idx, length(frame))] = v
                            end
                        end
                    end
                end
            end
        end
        push!(frames, __pytra_bytes(frame))
    end
    save_gif(out_path, w, h, frames, color_palette(), 3, 0)
    elapsed = (perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", frames_n)
    __pytra_print("elapsed_sec:", elapsed)
end


run_11_lissajous_particles();
